### DISCUSSION TITLE
Existing aggregate function on PostgreSQL

### DISCUSSION THREAD

**Discussion by mauricioszabo** [geldata/gel#2532] [Q&A]
Hi, just asking: is there a way to use already existing PostgreSQL functions, for example, `avg`, `string_agg`, etc?

**Comment by 1st1** [comment]
> Hi, just asking: is there a way to use already existing PostgreSQL functions

Unfortunately no, we have to expose them in our stdlib. Any other aggregates you'd want to see beyond the ones you mentioned?

> avg

We have it as https://www.edgedb.com/docs/edgeql/funcops/math#function::math::mean

> string_agg

You can workaround with https://www.edgedb.com/docs/edgeql/funcops/array#function::std::array_join and https://www.edgedb.com/docs/edgeql/funcops/array#function::std::array_agg. Although we should probably add `string_agg` (@elprans, @vpetrovykh?)

**Reply by mauricioszabo** [reply]
I'm mostly asking because I'm evaluating if EdgeDB could be an "almost drop-in replacement" from PostgreSQL, so being able to use, if not all, most functions that PG already have would be a huge "bonus points" for my use-case.

But thanks anyway. Is there a plan to make a compatibility layer over these functions, or even add support for users make these layers?

**Reply by 1st1** [reply]
When we expose those functions we usually do that very carefully, fixing any loose SQL semantics we see. Take `math::mean` as an example. We map it to `avg`, but Postgres' `avg` returns `NULL` when the set is empty (yet `SELECT 1/0` raises an exception in Postgres).  EdgeDB fixes this to be consistent, so both `SELECT 1/0` and `SELECT math::mean({})` predictably raise an error).

So our approach here is to expose functionality carefully, ideally driven by user requests. So please feel free to suggest what specific functions you need and we'll see how/if we can expose them.
